### PR TITLE
Rework unsafe

### LINKED ISSUE #1: Document `unsafe` with `SAFETY` comments and warn about usage of `unsafe` in README
From [reddit](https://www.reddit.com/r/rust/comments/14jasc6/comment/jpkr35z/?utm_source=reddit&utm_medium=web2x&context=3):

> amarao_san
> I checked few places, and I see a lot of unsafe code which I have trouble to understand. May be I'm not that bright, but you don't have any SAFETY comments there proving to readers that it's safe to have unsafe in safe code. If you want your library to be really reviewed, you need to write proofs for each unsafe part. "SAFETY: We can do it because of this, and this, and it will uphold invariant which compiler can't check but we know because ..."

> kinoshitajona
> 1. unsafe blocks should have a Safety comment explaining how you uphold the invariants required for this unsafe code usage.
> 1. unsafe fn or traits should have a Safety comment explaining what invariants the user must uphold.

https://std-dev-guide.rust-lang.org/documentation/safety-comments.html

https://rust-lang.github.io/api-guidelines/documentation.html#c-failure

Furthermore, I think it's good to warn users in README about the usage of unsafe in the library, so they are able to make a decision on if they want to pull `unsafe` code as a dependency.

### PR DESCRIPTION
**Pull request by wyfo**
Fix #1

### CODE REVIEW

File: src/buffer/vec.rs [hunk 1]
```diff
@@ -1,58 +1,71 @@
...
 
+    #[inline]
     fn capacity(&self) -> usize {
         self.0.len()
     }
 
+    #[inline]
     unsafe fn slice(&mut self, range: Range<usize>) -> Self::Slice<'_> {
-        &mut *(&mut self.0[range] as *mut [UnsafeCell<MaybeUninit<T>>] as *mut [T])
+        // SAFETY: [UnsafeCell<MaybeUninit<T>>] has the same layout as [T]
```
Comment (simonsan) [inline src/buffer/vec.rs:27]:
```suggestion
// SAFETY: [Cell<MaybeUninit<T>>] has the same layout as [T]
```

File: src/buffer/array.rs [hunk 2]
```diff
@@ -1,52 +1,66 @@
...
 
+    #[inline]
     fn capacity(&self) -> usize {
         self.0.len()
     }
 
+    #[inline]
     unsafe fn slice(&mut self, range: Range<usize>) -> Self::Slice<'_> {
-        mem::transmute(&mut self.0[range])
+        // SAFETY: [UnsafeCell<MaybeUninit<T>>] has the same layout as [T]
```
Comment (simonsan) [inline src/buffer/array.rs:30]:
```suggestion
// SAFETY: [Cell<MaybeUninit<T>>] has the same layout as [T]
```

File: src/buffer/vec.rs (hunk 1 above)
Comment (wyfo) [reply src/buffer/vec.rs:27]: I was not sure if you would do a review, so I've merged it a little bit too soon. Thank you a lot!

Comment (wyfo) [reply src/buffer/vec.rs:27]: Done in #7

Comment (simonsan) [reply src/buffer/vec.rs:27]: > I was not sure if you would do a review, so I've merged it a little bit too soon. Thank you a lot!

don't worry, just looked a bit over it and it's the only thing that jumped to my eyes kind of

**Pull request merged**